packages/utils/src/joiner/graphq-ast.ts

```diff
@@ -0,0 +1,140 @@
+import { RemoteJoinerQuery } from "@medusajs/types"
+import {
+  ArgumentNode,
+  DocumentNode,
+  FieldNode,
+  OperationDefinitionNode,
+  SelectionSetNode,
+  parse,
+} from "graphql"
+
+interface Argument {
+  name: string
+  //type: string;
+  value?: any
+}
+
+interface Entity {
+  property: string
+  fields: string[]
+  args?: Argument[]
+}
+
+class GraphQLParser {
+  private ast: DocumentNode
+
+  constructor(input: string, private variables?: { [key: string]: any }) {
+    this.ast = parse(input)
+    this.variables = variables || {}
+  }
+
+  private parseValueNode(valueNode: any): any {
+    if (valueNode.kind === "Variable") {
+      const variableName = valueNode.name.value
+      return this.variables ? this.variables[variableName] : undefined
+    } else if (valueNode.kind === "ObjectValue") {
+      let obj = {}
+      for (const field of valueNode.fields) {
+        obj[field.name.value] = this.parseValueNode(field.value)
+      }
+      return obj
+    }
+
+    return valueNode.value
+  }
+
+  private parseArguments(
+    args: readonly ArgumentNode[]
+  ): Argument[] | undefined {
+    if (!args.length) {
+      return
+    }
+
+    return args.map((arg) => {
+      const value = this.parseValueNode(arg.value as any)
+
+      return {
+        name: arg.name.value,
+        // type: value.kind, // TODO: use it to validate the type of the argument
+        value: value,
+      }
+    })
+  }
+
+  private extractEntities(
+    node: SelectionSetNode,
+    parentName = "",
+    mainService = ""
+  ): Entity[] {
+    const entities: Entity[] = []
+
+    node.selections.forEach((selection) => {
+      if (selection.kind === "Field") {
+        const fieldNode = selection as FieldNode
+
+        if (fieldNode.selectionSet) {
+          const entityName = parentName
+            ? `${parentName}.${fieldNode.name.value}`
+            : fieldNode.name.value
+
+          const nestedEntity: Entity = {
+            property: entityName.replace(`${mainService}.`, ""),
+            fields: fieldNode.selectionSet.selections.map(
+              (field) => (field as FieldNode).name.value
+            ),
+            args: this.parseArguments(fieldNode.arguments!),
+          }
+
+          entities.push(nestedEntity)
+          entities.push(
+            ...this.extractEntities(
+              fieldNode.selectionSet,
+              entityName,
+              mainService
+            )
+          )
+        }
+      }
+    })
+
+    return entities
+  }
+
+  public parseQuery(): RemoteJoinerQuery {
+    const queryDefinition = this.ast.definitions.find(
+      (definition) => definition.kind === "OperationDefinition"
+    ) as OperationDefinitionNode
```

Comment:
```suggestion
    const queryDefinition: OperationDefinitionNode = this.ast.definitions.find(
      (definition) => definition.kind === "OperationDefinition"
    )
```

Casting shouldn't be necessary here or in other places in this file if i'm not mistaken.

Reply:
it says `OperationDefinitionNode` is not assinable to type `DefinitionNode` :/

Reply:
fair, I'll resolve this fella.